pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+func ConcurrencyStateHandler(logger *zap.SugaredLogger, h http.Handler) http.HandlerFunc {
+	logger.Info("Concurrency state tracking enabled")
+
+	type req struct {
+		w http.ResponseWriter
+		r *http.Request
+
+		done chan struct{}
+	}
+
+	reqCh := make(chan req)
+	doneCh := make(chan struct{})
```

Comment:
do we want to buffer it to the concurrency level?

Reply:
Added in last commit.

cc @julz for your thoughts here

Reply:
I _think_ the only difference here would be whether the goroutines can immediately exit or not after they send to done, but that may or may not make any real difference and it adds a bit of complexity. I'd probably start with a simple no-buffer channel and then justify something cleverer with a benchmark, but it's very possible @vagababov has spotted an angle I've missed!

Reply:
ok -- I'll revert back to unbuffered and then work on a benchmark (and if we need to go buffered, probably do that in a separate PR)

Reply:
I don't want us to start waiting on a parallel request when something happens to en existing one. But 🤷 , might be overthingking.

Reply:
Dunno, but I am not thrilled that arbitrary pause/resume funcs are called on the serving thread, but we have control, so I guess it's ok.